Boot mode checks

tescande opened the pull request.

tescande committed:
Ignore backup if boot modes don't match

This patch checks the boot mode of an existing backups and disallows its
restoration if it differs from the installer one. This uses the new
helper function is_rootf_uefi().

Signed-off-by: Thierry Escande <thierry.escande@vates.tech>

tescande committed:
Ignore existing installation for upgrade if boot modes don't match

This patch checks that the current installation is not proposed as
upgradeable if its boot mode differs from the installer one. This uses
the helper function is_rootfs_uefi().

Signed-off-by: Thierry Escande <thierry.escande@vates.tech>

ydirson commented on product.py:
```diff
@@ -63,6 +63,20 @@ def isUpgradeable(self):
+            if 'id' in boot_part and boot_part['id'] == GPTPartitionTool.ID_EFI_BOOT:
+                existing_boot_mode = constants.TARGET_BOOT_MODE_UEFI
+
+            if (existing_boot_mode == constants.TARGET_BOOT_MODE_UEFI and not constants.UEFI_INSTALLER) or \
+                (existing_boot_mode == constants.TARGET_BOOT_MODE_LEGACY and constants.UEFI_INSTALLER):
```
a bit heavy to read, what about something like this (closer to XOR)
```
if (existing_boot_mode == constants.TARGET_BOOT_MODE_UEFI) != constants.UEFI_INSTALLER:
```

Will also allow to get a visually clear `if` statement without a double indent (never seen that in the wild, pretty sure the code checkers would flag it)

tescande replied:
Done

ydirson commented on product.py:
```diff
@@ -566,6 +580,22 @@ def findXenSourceBackups():
+                    if m:
+                        backup_is_uefi = True
+                        break
+            if (backup_is_uefi and not constants.UEFI_INSTALLER) or \
+                (not backup_is_uefi and constants.UEFI_INSTALLER):
```
XOR too:
```
if backup_is_uefi != constants.UEFI_INSTALLER):
```

ydirson commented on product.py:
```diff
@@ -566,6 +580,22 @@ def findXenSourceBackups():
+                        break
+            if (backup_is_uefi and not constants.UEFI_INSTALLER) or \
+                (not backup_is_uefi and constants.UEFI_INSTALLER):
+                logger.log("findXenSourceBackups: ignoring, installer mode (%s) is "
+                            "mismatched with bakcup boot mode (%s)" %
```
"does not match"? + "bakcup" typo

ydirson commented on product.py:
```diff
@@ -566,6 +580,22 @@ def findXenSourceBackups():
+            backup_is_uefi = False
+            with open(os.path.join(b.mount_point, 'etc', 'fstab'), 'r') as fstab:
+                for line in fstab:
+                    m = re.search("/boot/efi[[:space:]]+vfat", line)
```
do we really need to check the fstype?
I suggest something like `if line.split(maxsplit=1)[0] == "/boot/efi"`

tescande replied:
Actually changed it to `re.search(r'^[^#].*\s+/boot/efi', line.strip())` to ignore commented lines

ydirson replied:
still, a regex seems overkill: we only need to check for the first word, and `split()` will even take care of any leading whitespace

tescande replied:
As discussed, we'll stick with a `re.search()`

ydirson commented on product.py:
```diff
@@ -63,6 +63,20 @@ def isUpgradeable(self):
+                (existing_boot_mode == constants.TARGET_BOOT_MODE_LEGACY and constants.UEFI_INSTALLER):
+                    logger.log("Cannot upgrade %s, installer mode (%s) is mismatched with existing boot mode (%s)" %
+                                (self.primary_disk, "UEFI" if constants.UEFI_INSTALLER else "legacy", existing_boot_mode))
+                    raise Exception("Boot mode mismatch")
```
`return False` would me more direct, this descriptive string is not used

tescande replied:
Done

ydirson commented on product.py:
```diff
@@ -63,6 +63,20 @@ def isUpgradeable(self):
         try:
+            # Don't propose to upgrade a BIOS installation with a UEFI installer and conversely
+            tool = PartitionTool(self.primary_disk)
+            existing_boot_mode = constants.TARGET_BOOT_MODE_LEGACY
```
moving this to an `else` arm would be more straightforward

tescande replied:
Done

ydirson reviewed (changes requested):
Do we really want to check at upgrade time and restore time using different criteria?  What about putting the logic in a helper function?

ydirson commented on product.py:
```diff
@@ -566,6 +566,19 @@ def findXenSourceBackups():
             logger.log("Found a backup: %s" % (repr(backup),))
 
+            # Don't restore a BIOS backup with a UEFI installer and conversely
+            backup_is_uefi = False
```
nitpicking and possibly a matter of taste: that one would fit neatly in an `else` arm of the `for` loop

tescande replied:
TIL `for: else:` loop.
Fixed, thanks

tescande committed:
Add helper function is_rootfs_uefi()

This function searches for the /boot/efi mount point in the /etc/fstab
file of the mount point directory passed as parameter and returns True
if present, False otherwise.

This will allow to check that an existing installation or backup boot
mode matches with the installer boot mode.

Signed-off-by: Thierry Escande <thierry.escande@vates.tech>

tescande commented:
> Do we really want to check at upgrade time and restore time using different criteria? What about putting the logic in a helper function?

You're right so I ended up using the `/boot/efi` entry in `/etc/fstab` for both checks.

tescande commented on product.py:
```diff
@@ -28,6 +28,18 @@ class SettingsNotAvailable(Exception):
+                m = re.search(r'^[^#].*\s+/boot/efi\s+', line.strip())
+                if m:
+                    return True
+    except:
+        pass
```
Ok, will catch `FileNotFoundError`

tescande commented on product.py:
```diff
@@ -28,6 +28,18 @@ class SettingsNotAvailable(Exception):
+    try:
+        with open(os.path.join(mount_point, 'etc', 'fstab'), 'r') as fstab:
+            for line in fstab:
+                m = re.search(r'^[^#].*\s+/boot/efi\s+', line.strip())
```
Ok, will do it on the regex instead
